### ISSUE TITLE
Results between multi:softprob and custom softmax do not match

### ISSUE THREAD

**Issue by rydevera3** [dmlc/xgboost#4467]
I am trying to create a custom softmax objective for multiclass classification in xgboost. To start I thought I would start with the basic softmax case. I use `multi:softprob` and a custom function. Here is a reproducible example:

```
import numpy as np
from sklearn import datasets
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
import lightgbm as lgb
from sklearn.metrics import log_loss
import xgboost as xgb

# Iris dataset
iris = datasets.load_iris()
X, Ymc = iris.data, iris.target

X_train, X_test, y_train, y_test = train_test_split(X, Ymc, test_size=0.30, random_state=2019)

def softmaxobj(preds, dtrain):
    
    labels = dtrain.get_label()
    labels = labels.reshape(-1, 1)
    labels = OneHotEncoder(sparse=False, categories='auto').fit_transform(labels)
    
    grad = preds - labels
    hess = 2.0 * preds * (1.0 - preds)
    return grad.flatten('F'), hess.flatten('F')

dtrain = xgb.DMatrix(X_train, label=y_train)
dvalid = xgb.DMatrix(X_test, label=y_test)
dtest = xgb.DMatrix(X_test)
eval_set=[(dvalid, 'eval')]

param = {'max_depth': '6', 'objective':'multi:softprob', 
         'tree_method':'hist', 'num_class': 3,
         'eval_metric': 'mlogloss'}

bst = xgb.train(param, dtrain, num_boost_round=100, 
                verbose_eval=10, evals=eval_set)

preds=bst.predict(dtest)
print(log_loss(y_test, preds))

[0]	eval-mlogloss:0.749971
[10]	eval-mlogloss:0.111583
[20]	eval-mlogloss:0.107522
[30]	eval-mlogloss:0.116207
[40]	eval-mlogloss:0.119493
[50]	eval-mlogloss:0.125152
[60]	eval-mlogloss:0.130457
[70]	eval-mlogloss:0.134239
[80]	eval-mlogloss:0.138156
[90]	eval-mlogloss:0.141714
[99]	eval-mlogloss:0.144838
custom objective logloss0.14483839407071677

#train xgboost with same custom loss function
param = {'max_depth': '6', 'objective':'multi:softprob', 
         'tree_method':'hist', 'num_class': 3,
         'eval_metric': 'mlogloss'}

bst_cust = xgb.train(param, dtrain, num_boost_round=100, 
                     verbose_eval=10, evals=eval_set,
                     obj=softmaxobj)

preds_cust=bst_cust.predict(dtest)
print('custom objective {}'.format(log_loss(y_test, preds_cust)))

[0]	eval-mlogloss:1.16279
[10]	eval-mlogloss:2.63613
[20]	eval-mlogloss:5.60595
[30]	eval-mlogloss:10.83
[40]	eval-mlogloss:15.2783
[50]	eval-mlogloss:17.3254
[60]	eval-mlogloss:18.8124
[70]	eval-mlogloss:19.3257
[80]	eval-mlogloss:18.956
[90]	eval-mlogloss:17.9979
[99]	eval-mlogloss:17.567
custom objective 16.74834724995825
```

I tried to follow the example given here - https://gist.github.com/ytsaig/a596feec53d2a024ac69f5ae5a83d8f7 but I am unsuccessful. Has anyone been able to get successful results with a custom multiclass objective function?

**Comment by trivialfis** [comment]
@rydevera3 Since the training error is increasing, I would say your implementation is incorrect.  I think you forgot to compute `softmax(preds)` to transform preds first.

**Comment by trivialfis** [comment]
@rydevera3 Closing for now.  Please verify your implementation of `softmax` objective.  Feel free to reopen if the result is still different after adding correct softmax transformation.
